core/block/detailservice/set_details.go

```diff
@@ -0,0 +1,248 @@
+	objectIDsPerSpace, err := s.partitionObjectIDsBySpaceID(objectIds)
+	if err != nil {
+		return fmt.Errorf("partition object ids by spaces: %w", err)
+	}
+
+	var (
+		multiErr   multierror.Error
+		anySucceed bool
+	)
+	for spaceID, objectIDs := range objectIDsPerSpace {
+		err = s.setIsArchivedForObjects(spaceID, objectIDs, isArchived)
+		if err != nil {
+			log.Error("failed to set isArchived to objects", zap.String("spaceId", spaceID),
+				zap.Strings("objectIDs", objectIDs), zap.Bool("isArchived", isArchived), zap.Error(err))
+			multiErr.Errors = append(multiErr.Errors, err)
+		} else {
+			anySucceed = true
+		}
+	}
+	if anySucceed {
+		return nil
+	}
+	return multiErr.ErrorOrNil()
+}
+
+func (s *service) checkArchivedRestriction(isArchived bool, objectId string) error {
+	if !isArchived {
+		return nil
+	}
+	return cache.Do(s, objectId, func(sb smartblock.SmartBlock) error {
+		return s.restriction.CheckRestrictions(sb, model.Restrictions_Delete)
+	})
+}
+
+func (s *service) objectLinksCollectionModify(collectionId string, objectId string, value bool) error {
+	return cache.Do(s, collectionId, func(b smartblock.SmartBlock) error {
+		coll, ok := b.(collection.Collection)
+		if !ok {
+			return fmt.Errorf("unsupported sb block type: %T", b)
+		}
+		if value {
+			return coll.AddObject(objectId)
+		} else {
+			return coll.RemoveObject(objectId)
+		}
+	})
+}
+
+func (s *service) partitionObjectIDsBySpaceID(objectIDs []string) (map[string][]string, error) {
+	res := map[string][]string{}
+	for _, objectID := range objectIDs {
+		spaceID, err := s.resolver.ResolveSpaceID(objectID)
+		if err != nil {
+			return nil, fmt.Errorf("resolve spaceID: %w", err)
+		}
+		res[spaceID] = append(res[spaceID], objectID)
+	}
+	return res, nil
+}
+
+func (s *service) setIsArchivedForObjects(spaceID string, objectIDs []string, isArchived bool) error {
+	spc, err := s.spaceService.Get(context.Background(), spaceID)
+	if err != nil {
+		return fmt.Errorf("get space: %w", err)
+	}
+	return cache.Do(s, spc.DerivedIDs().Archive, func(b smartblock.SmartBlock) error {
+		archive, ok := b.(collection.Collection)
+		if !ok {
+			return fmt.Errorf("unexpected archive block type: %T", b)
+		}
+
+		var multiErr multierror.Error
+		var anySucceed bool
+		ids, err := s.store.HasIDs(objectIDs...)
+		if err != nil {
+			return err
+		}
+		for _, id := range ids {
+			var err error
+			if restrErr := s.checkArchivedRestriction(isArchived, id); restrErr != nil {
+				err = restrErr
+			} else {
+				if isArchived {
+					err = archive.AddObject(id)
+				} else {
+					err = archive.RemoveObject(id)
+				}
+			}
+			if err != nil {
+				log.Error("failed to set isArchived", zap.String("objectId", id), zap.Bool("isArchived", isArchived), zap.Error(err))
+				multiErr.Errors = append(multiErr.Errors, err)
+				continue
+			}
+			anySucceed = true
+		}
+
+		if err := multiErr.ErrorOrNil(); err != nil {
+			log.Warn("failed to archive", zap.Error(err))
+		}
+		if anySucceed {
```

Comment:
It is seems like unnecessary check, because multiErr.ErrorOrNil() is nil for sure. So can we just return nil instead of lines 243-246?

Reply:
The point here is to return nil, if at least one object was processed without error.
If some objects were processed with error, than we should log it.
If all objects were processed with error, than we should return error.
So there is no issue here